### PR TITLE
Add Speaker Diarization and Verification heads

### PR DESCRIPTION
**Pull request by anton-l**
# What does this PR do?

This adds the Audio Frame classification (equivalent of token classification) and X-vector (speaker embedding extraction) heads to `Wav2Vec2` and `UniSpeech-SAT` models. The target tasks for the heads are SUPERB's Speaker Diarization and Speaker Verification respectively. These were mainly motivated by `UniSpeech-SAT`, since the model performs better on those tasks, rather than on ASR.

Sources for the models and weights from the SUPERB's `s3prl` framework:
* `ModelForAudioFrameClassification`: https://github.com/s3prl/s3prl/blob/master/s3prl/downstream/diarization/model.py (with `use_rnn=False`)
* `ModelForXVector`: https://github.com/s3prl/s3prl/blob/master/s3prl/downstream/sv_voxceleb1/model.py#L261

The heads for both W2V2 and US-SAT were finetuned from scratch, since the official checkpoints use custom (better) heads that are incompatible with SUPERB's evaluation protocol.

### CODE REVIEW

File: src/transformers/models/unispeech/modeling_unispeech.py [hunk 1]
```diff
@@ -1525,3 +1560,319 @@ def forward(
...
+            norm_weights = nn.functional.softmax(self.layer_weights, dim=-1)
+            hidden_states = (hidden_states * norm_weights.view(-1, 1, 1)).sum(dim=1)
+        else:
+            hidden_states = outputs[0]
+
+        logits = self.classifier(hidden_states)
+
+        loss = None
+        if labels is not None:
+            loss = self.pit_loss(logits, attention_mask, labels)
```
Comment (patrickvonplaten) [inline src/transformers/models/unispeech/modeling_unispeech.py:1667]: Think we don't even need the loss function in a first step as we'll not provide an example script here

Comment (patrickvonplaten) [reply src/transformers/models/unispeech/modeling_unispeech.py:1667]: But if it works already correctly - happy to leave it

**Comment by anton-l** [comment]
@patrickvonplaten the implementation is ready for a full review now. I'll post the model links and evaluation results (DER and EER metrics) shortly.

**Comment by patrickvonplaten** [comment]
checking tomorrow after the `-large` checkpoint problem is solved :-)

File: docs/source/model_doc/auto.rst [hunk 2]
```diff
@@ -195,6 +202,13 @@ AutoModelForSpeechSeq2Seq
     :members:
 
 
+AutoModelForAudioXVector
```
Comment (patrickvonplaten) [inline docs/source/model_doc/auto.rst:205]: like the two AutoNames as they relate to the architecture

File: src/transformers/file_utils.py [hunk 3]
```diff
@@ -1117,6 +1117,51 @@ def _prepare_output_docstrings(output_type, config_class):
...
+        >>> model = {model_class}.from_pretrained('{checkpoint}')
+
+        >>> # audio file is decoded on the fly
+        >>> inputs = feature_extractor(dataset[:2]["audio"]["array"], return_tensors="pt")
+        >>> embeddings = model(**inputs).embeddings
+        >>> embeddings = torch.nn.functional.normalize(embeddings, dim=-1).cpu()
+
+        >>> # the resulting embeddings can be used for cosine similarity-based retrieval
+        >>> cosine_sim = torch.nn.CosineSimilarity(dim=-1)
+        >>> similarity = cosine_sim(embeddings[0], embeddings[1])
```
Comment (patrickvonplaten) [inline src/transformers/file_utils.py:1162]: Very nice!

Could you maybe add one more line that shows over which similarity the "verification" would be True? E.g.:
```python
assert similarity > 0.5 "Speakers are not the same!"
```

File: src/transformers/models/unispeech_sat/configuration_unispeech_sat.py [hunk 4]
```diff
@@ -247,6 +262,10 @@ def __init__(
         self.do_stable_layer_norm = do_stable_layer_norm
         self.use_weighted_layer_sum = use_weighted_layer_sum
         self.classifier_proj_size = classifier_proj_size
+        self.tdnn_dim = list(tdnn_dim)
```
Comment (patrickvonplaten) [inline src/transformers/models/unispeech_sat/configuration_unispeech_sat.py:265]: (nit) could you maybe put this under a comment like:

```# xvector-specific configs. Feel free to ignore for other classes```

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 5]
```diff
@@ -1773,3 +1813,282 @@ def forward(
...
+
+class TDNNLayer(nn.Module):
+    def __init__(self, config, layer_id=0):
+        super().__init__()
+        self.in_conv_dim = config.tdnn_dim[layer_id - 1] if layer_id > 0 else config.tdnn_dim[layer_id]
+        self.out_conv_dim = config.tdnn_dim[layer_id]
+        self.kernel_size = config.tdnn_kernel[layer_id]
+        self.dilation = config.tdnn_dilation[layer_id]
+
+        # TODO: replace with Conv1D
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1938]: what is the TODO for?

Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:1938]: can be removed?

File: src/transformers/models/unispeech_sat/configuration_unispeech_sat.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/transformers/models/unispeech_sat/configuration_unispeech_sat.py:265]: applies to all other configs as well

**Review by patrickvonplaten — APPROVED** [review]
Very clean! Only left a couple of nits

File: src/transformers/models/wav2vec2/convert_wav2vec2_original_s3prl_checkpoint_to_pytorch.py [hunk 6]
```diff
@@ -19,13 +19,57 @@
...
+
+def convert_xvector(base_model_name, hf_config, downstream_dict):
+    model = Wav2Vec2ForXVector.from_pretrained(base_model_name, config=hf_config)
+    model.projector.weight.data = downstream_dict["connector.weight"]
+    model.projector.bias.data = downstream_dict["connector.bias"]
+    for i, kernel_size in enumerate(hf_config.tdnn_kernel):
+        # kernel = torch.tensor(downstream_dict[f"model.framelevel_feature_extractor.module.{i}.kernel.weight"])
+        # prev_dim = hf_config.tdnn_dim[i - 1] if i > 0 else hf_config.tdnn_dim[i]
+        # model.tdnn[i].conv.weight.data = kernel.reshape(prev_dim, kernel_size, hf_config.tdnn_dim[i]).permute(2, 0, 1)
+        # model.tdnn[i].conv.bias.data = downstream_dict[f"model.framelevel_feature_extractor.module.{i}.kernel.bias"]
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/convert_wav2vec2_original_s3prl_checkpoint_to_pytorch.py:60]: To clean up?

File: utils/update_metadata.py [hunk 7]
```diff
@@ -74,6 +74,12 @@
         "MODEL_FOR_NEXT_SENTENCE_PREDICTION_MAPPING_NAMES",
         "AutoModelForNextSentencePrediction",
     ),
+    (
+        "audio-frame-classification",
+        "MODEL_FOR_AUDIO_FRAME_CLASSIFICATION_MAPPING_NAMES",
+        "AutoModelForAudioFrameClassification",
+    ),
+    ("audio-xvector", "MODEL_FOR_AUDIO_XVECTOR_MAPPING_NAMES", "AutoModelForAudioXVector"),
```
Comment (sgugger) [inline utils/update_metadata.py:82]: Nice that you didn't forget to add them here!

**Review by sgugger — APPROVED** [review]
Very clean, thanks for adding those!

**Review by LysandreJik — APPROVED** [review]
Overall very cool addition! Just wondering a bit about the `XVector` but if that's a very widely used term in audio, fine for me!

File: docs/source/model_doc/auto.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (anton-l) [reply docs/source/model_doc/auto.rst:205]: ForAudioXVector pretty much summarizes "for extracting speaker embeddings from a model that was trained as a classifier", so @patrickvonplaten and I opted to keep it :)
